components/https_server/src/https_server.c

```diff
@@ -0,0 +1,203 @@
+#include <string.h>
+#include "https_server.h"
+#include "openssl/ssl.h"
+#include "esp_log.h"
+#include "sdkconfig.h"
+
+const static char *TAG = "https_server";
+
+/**
+ * SSL socket close handler (ctx_free session attribute)
+ *
+ * @param ctx
+ */
+static void httpdssl_close(void *ctx)
+{
+    assert(ctx != NULL);
+    SSL_shutdown(ctx);
+    SSL_free(ctx);
+    ESP_LOGD(TAG, "Secure socket closed");
+}
+
+/**
+ * SSL socket pending-check function
+ *
+ * @param server
+ * @param sockfd
+ * @return number of pending bytes, negative on error
+ */
+static int httpdssl_pending(httpd_handle_t server, int sockfd)
+{
+    SSL *ssl = httpd_sess_get_ctx(server, sockfd);
+    assert(ssl != NULL);
+    return SSL_pending(ssl);
+}
+
+/**
+ * Receive from a SSL socket
+ *
+ * @param server
+ * @param sockfd
+ * @param buf
+ * @param buf_len
+ * @param flags
+ * @return bytes read, negative on error
+ */
+static int httpdssl_recv(httpd_handle_t server, int sockfd, char *buf, size_t buf_len, int flags)
+{
+    SSL *ssl = httpd_sess_get_ctx(server, sockfd);
```

Comment:
Session context is passed along with the httpd_req_t structure when URI handler is executed. It's useful for session persistence (See the `persistent_sockets` example).

Could you use a different member variable (and corresponding APIs) for saving and accessing the pointer to SSL object. Maybe call it `priv_ctx` or `internal_ctx`?

Reply:
sure. in that case, do you want to keep the functions `httpd_sess_get_ctx` / `httpd_sess_set_ctx` that I added for this purpose as well as adding new ones for this 'internal_ctx'?

Reply:
You can keep them. They may come handy sometime.